### PR TITLE
BlockUndo finalise work to make it a map instead of vec

### PR DESCRIPTION
**Pull request by azarovh**
This allows to undo an arbitrary tx if no other tx depends on it

### CODE REVIEW

File: utxo/src/undo.rs [hunk 1]
```diff
@@ -100,20 +110,47 @@ impl BlockUndo {
...
+            // If not this tx can be taken and returned.
+            // But first, remove itself as a dependency of others.
+            self.dependencies
+                .iter()
+                .filter_map(|v| if v.0 == *tx_id { Some(*v) } else { None })
+                .collect::<Vec<_>>()
+                .iter()
+                .for_each(|p| {
+                    self.dependencies.remove(p);
+                });
```
Comment (stanislav-tkach) [inline utxo/src/undo.rs:148]:
```suggestion
            self.dependencies.retain(|v| &v.0 == tx_id);
```

**Review by stanislav-tkach — APPROVED** [review]

File: utxo/src/undo.rs [hunk 2]
```diff
@@ -100,20 +110,47 @@ impl BlockUndo {
...
-        // TODO(PR): this should analyze dependencies before taking an undo
-        self.tx_undos.remove(tx_id)
+        // Check if the tx is a dependency for other txs.
+        let dependencies_count =
+            self.dependencies.iter().filter(|(_child, parent)| parent == tx_id).count();
+        if dependencies_count == 0 {
+            // If not this tx can be taken and returned.
+            // But first, remove itself as a dependency of others.
+            self.dependencies
+                .iter()
```
Comment (TheQuantumPhysicist) [inline utxo/src/undo.rs:142]: Unfortunately, this is a bad design. This being O(N) basically means that undoing blocks is potentially O(N^2). That's a recipe for disaster with DoS attacks. We can't accept that. 

This should be at most O(log(N)). Meaning: You should provide a mechanism to search for transactions that doesn't involve looping over all transactions, but use some kind of tree at worst (like BTreeSet/Map).

File: utxo/src/undo.rs [hunk 3]
```diff
@@ -100,20 +110,47 @@ impl BlockUndo {
...
+
         Ok(())
     }
 
     pub fn take_tx_undo(&mut self, tx_id: &Id<Transaction>) -> Option<TxUndo> {
-        // TODO(PR): this should analyze dependencies before taking an undo
-        self.tx_undos.remove(tx_id)
+        // Check if the tx is a dependency for other txs.
+        let dependencies_count =
+            self.dependencies.iter().filter(|(_child, parent)| parent == tx_id).count();
```
Comment (TheQuantumPhysicist) [inline utxo/src/undo.rs:137]: This is also bad for the same reason. You should never use iter() in this place. You should using something like `self.dependencies.range()`, but it seems that the dependencies are inverted. I expected this to be the other way around.

File: utxo/src/undo.rs [hunk 4]
```diff
@@ -100,20 +118,52 @@ impl BlockUndo {
...
+        let dependencies_count =
+            self.parent_child_dependencies.range(range_start..=range_end).count();
+        if dependencies_count == 0 {
+            // If not this tx can be taken and returned.
+            // But first, remove itself as a dependency of others.
+            let to_remove = self
+                .child_parent_dependencies
+                .range(range_start..=range_end)
+                .copied()
+                .collect::<Vec<_>>();
```
Comment (TheQuantumPhysicist) [inline utxo/src/undo.rs:157]: I don't think you need to collect here... you could just use the iterator in the next step.

Comment (azarovh) [reply utxo/src/undo.rs:157]: Actually, I do, because otherwise, I would hold references to `child_parent_dependencies` and won't be able to modify it

Comment (TheQuantumPhysicist) [reply utxo/src/undo.rs:157]: I see.

Comment (stanislav-tkach) [reply utxo/src/undo.rs:157]: Perhaps the `drain_filter` method can be used here, but it is still unstable.

**Review by stanislav-tkach — APPROVED** [review]

**Pull request merged**